cnc_ctrl_v1/Calibration.cpp

```diff
@@ -0,0 +1,46 @@
+/*This file is part of the Maslow Control Software.
+
+The Maslow Control Software is free software: you can redistribute it and/or modify
+it under the terms of the GNU General Public License as published by
+the Free Software Foundation, either version 3 of the License, or
+(at your option) any later version.
+
+Maslow Control Software is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU General Public License for more details.
+
+You should have received a copy of the GNU General Public License
+along with the Maslow Control Software.  If not, see <http://www.gnu.org/licenses/>.
+
+Copyright 2014-2017 Bar Smith*/
+
+#include "Maslow.h"
+#include <EEPROM.h>
+
+void initializeCalibration() {
+  Serial.print(F("Zeroing Calibration\r\n"));
+  for (int x=0; x<31; x++) {
+    for (int y=0; y<15; y++) {
+      calibration.xError[x][y] = 0;
+      calibration.yError[x][y] = 0;
+    }
+  }
+}
+
+byte calibrationUpdateMatrix(int x, int y, int xValue, int yValue) {
+  if ((x==31) && (y==15)) {
+    settingsSaveToEEprom();
+    return(STATUS_OK);
+  } else {
+    if (((x>=0) && (x<31)) && ((y>=0) && (y<15))) {
+      calibration.xError[x][y] = xValue;
+      calibration.yError[x][y] = yValue;
+      if ((x==15) && (y==7)) {
+        kinematics.init();
+      }
+      return(STATUS_OK);
+    }
+    return(STATUS_INVALID_STATEMENT);
```

Comment:
Might be simpler to write this as
```C
byte calibrationUpdateMatrix(int x, int y, int xValue, int yValue) {
  if ((x == 31) && (y == 15)) {
    settingsSaveToEEprom();
    return(STATUS_OK);
  } else if ((x < 0) || (x >= 31) || (y < 0) && (y >= 15)) {
    return(STATUS_INVALID_STATEMENT);
  }
  calibration.xError[x][y] = xValue;
  calibration.yError[x][y] = yValue;
  if ((x==15) && (y==7)) {
    kinematics.init();
  }
  return(STATUS_OK);
}
```

Reply:
I wasn't able to go through and confirm anything in detail.  We might call it enableXYAdjustment, rather than enableOpticalCalibration.  The (X,Y) adjustments could be determined via a bunch of mechanisms besides optical calibration.  In triangularInverse, it looks like you verifyValidTarget twice: once before adjustment and once after.  Is that necessary?  Would this be better implemented as an if/else, where the first verifyValidTarget is in moved into the else block?

Reply:
The naming change is a good idea. 

For `verifyValidTarget` it’s a good question. I’ll take a look again and see if this can get cleaned up.